File: apps/ios/Core/GenerationEpoch.swift
import Foundation

final class GenerationEpoch {
    private var current: Int64 = 0

    func next() -> Int64 {
        current += 1
        return current
    }

    func cancel() -> Int64 {
        current += 1
        return current
    }

    func isCurrent(_ generationId: Int64) -> Bool {
        current == generationId
    }
}

File: apps/android/app/src/main/kotlin/com/conversationalai/agent/core/GenerationEpoch.kt
package com.conversationalai.agent.core

import java.util.concurrent.atomic.AtomicLong

/** Monotonic cancellation epoch for dropping stale turn work after stop/barge-in/new turns. */
class GenerationEpoch(initialValue: Long = 0L) {
    private val current = AtomicLong(initialValue)

    fun next(): Long = current.incrementAndGet()

    fun cancel(): Long = current.incrementAndGet()

    fun isCurrent(generationId: Long): Boolean = generationId == current.get()
}
